devito/passes/iet/definitions.py

```diff
@@ -74,41 +79,42 @@ def _alloc_object_on_low_lat_mem(self, site, obj, storage):
         """
         Allocate a LocalObject in the low latency memory.
         """
-        storage.update(obj, site, allocs=c.Value(obj._C_typename, obj.name))
+        decl = Definition(obj, constructor_args=obj.constructor_args)
+
+        storage.update(obj, site, objs=decl)
 
     def _alloc_array_on_low_lat_mem(self, site, obj, storage):
         """
         Allocate an Array in the low latency memory.
         """
         shape = "".join("[%s]" % ccode(i) for i in obj.symbolic_shape)
         alignment = self.lang['aligned'](obj._data_alignment)
-        decl = c.Value(obj._C_typedata, "%s%s %s" % (obj._C_name, shape, alignment))
-
-        if obj.initvalue is not None:
-            storage.update(obj, site,
-                           allocs=c.Initializer(decl, ListInitializer(obj.initvalue)))
+        if obj.initvalue is None:
+            initvalue = None
         else:
-            storage.update(obj, site, allocs=decl)
+            initvalue = ListInitializer(obj.initvalue)
+        alloc = Definition(obj, shape=shape, qualifier=alignment, initvalue=initvalue)
+
+        storage.update(obj, site, allocs=alloc)
 
     def _alloc_scalar_on_low_lat_mem(self, site, expr, storage):
         """
         Allocate a Scalar in the low latency memory.
         """
-        key = (site, expr.write)  # Ensure a scalar isn't redeclared in the given site
-        storage.map(key, expr, expr._rebuild(init=True))
+        storage.map(expr.write, site, expr, expr._rebuild(init=True))
 
     def _alloc_array_on_high_bw_mem(self, site, obj, storage, *args):
         """
         Allocate an Array in the high bandwidth memory.
         """
-        decl = c.Value(obj._C_typedata, "*%s" % obj._C_name)
+        decl = Definition(obj)
 
-        shape = "".join("[%s]" % i for i in obj.symbolic_shape)
-        size = "sizeof(%s%s)" % (obj._C_typedata, shape)
-        alloc = c.Statement(self.lang['alloc-host'](obj._C_name,
-                                                    obj._data_alignment, size))
+        memptr = VOID(Byref(obj._C_symbol), '**')
```

Comment:
Note for reviewers: at last (!) these aren't cgen objects anymore, but rather first-class IET nodes. So now the visitors pick them up...